ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/reading/StateReader.scala

```diff
@@ -12,12 +12,43 @@ import scala.concurrent.{ExecutionContext, Future}
   * @tparam Value The type of the value returned.
   */
 trait StateReader[Key, Value] {
+  self =>
 
   /**
     * Reads values of a set of keys from the backing store.
     *
     * @param keys list of keys to look up
     * @return values corresponding to the requested keys, in the same order as requested
     */
-  def read(keys: Seq[Key])(implicit executionContext: ExecutionContext): Future[Seq[Option[Value]]]
+  def read(keys: Seq[Key])(implicit executionContext: ExecutionContext): Future[Seq[Value]]
+
+  /**
+    * Create a new StateReader that transforms the keys before reading.
+    *
+    * @param f the transformation function
+    * @tparam NewKey the type of the keys
+    * @return a `StateReader[NewKey, Value]`
+    */
+  def comapKeys[NewKey](f: NewKey => Key): StateReader[NewKey, Value] =
```

Comment:
Why `comap` rather than `map`?

Reply:
Because the mapping is backwards. `map` would take an `f: Key => NewKey`, but this takes `f: NewKey => Key`.

However, upon researching this a bit more, it seems I confused cofunctors with [contravariant functors](https://hackage.haskell.org/package/base/docs/Data-Functor-Contravariant.html). So this should probably be called `contramapKeys`.

Reply:
`reverseMapKeys`?

Reply:
(or `unmapKeys`)

Reply:
I don't think either of those conveys the right information. "Reverse" implies to me that the sequence of keys is reversed, and "unmap" makes me think I'm undoing a previous mapping operation.

While I agree with you that these functional programming terms are somewhat obtuse, we already use them _a lot_ elsewhere and they have definitions [which are fairly easy to search for](https://duckduckgo.com/?q=contramap). So I've settled for trying to explain it in a brief comment.